src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs

```diff
@@ -0,0 +1,326 @@
+            out SemanticVersion lastUpdateVersion,
+            out DateTime lastUpdateDate)
+        {
+            updateFilePath = null;
+            lastUpdateVersion = null;
+            lastUpdateDate = default;
+
+            var files = Directory.EnumerateFiles(s_cacheDirectory, UpdateFileNamePattern, s_enumOptions);
+            var enumerator = files.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                // No file was found that matches the pattern, but it's OK that an update file doesn't exist.
+                // This could happen when there is no new updates yet.
+                return true;
+            }
+
+            updateFilePath = enumerator.Current;
+            if (enumerator.MoveNext())
+            {
+                // More than 1 files were found that match the pattern. This is a corrupted state.
+                // Theoretically, there should be only one update file at any point of time.
+                updateFilePath = null;
+                return false;
+            }
+
+            // OK, only found one update file, which is expected.
+            // Now let's parse the file name.
+            string updateFileName = Path.GetFileName(updateFilePath);
+            int dateStartIndex = updateFileName.LastIndexOf('_') + 1;
+
+            bool success = DateTime.TryParse(
+                updateFileName.AsSpan().Slice(dateStartIndex),
+                CultureInfo.InvariantCulture,
+                DateTimeStyles.AssumeLocal,
+                out lastUpdateDate);
+
+            if (success)
+            {
+                int versionStartIndex = updateFileName.IndexOf('_') + 2;
+                int versionLength = dateStartIndex - versionStartIndex - 1;
+                string versionString = updateFileName.Substring(versionStartIndex, versionLength);
+                success = SemanticVersion.TryParse(versionString, out lastUpdateVersion);
+            }
+
+            if (!success)
+            {
+                updateFilePath = null;
+                lastUpdateVersion = null;
+                lastUpdateDate = default;
+            }
+
+            return success;
+        }
+
+        private static async Task<Release> QueryNewReleaseAsync(SemanticVersion baselineVersion)
+        {
+            bool notPreRelease = string.IsNullOrEmpty(PSVersionInfo.PSCurrentVersion.PreReleaseLabel);
+            string queryUri = notPreRelease ? LatestReleaseUri : Last3ReleasesUri;
+
+            using (HttpClient client = new HttpClient())
+            {
+                string userAgent = string.Format(CultureInfo.InvariantCulture, "PowerShell {0}", PSVersionInfo.GitCommitId);
+                client.DefaultRequestHeaders.Add("User-Agent", userAgent);
+                client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+
+                var response = await client.GetAsync(queryUri);
+                response.EnsureSuccessStatusCode();
```

Comment:
Does this throw if the response was not success? If so, might be worth a comment indicating that that's intended and handled